PowerNSX.psm1

```diff
@@ -27578,6 +27578,383 @@ function Get-NsxSecurityPolicy {
+        [ValidateSet("true","false")]
+            [boolean]$Logging = "false",
+        [Parameter (Mandatory=$false)]
+        [ValidateSet("allow","block", "reject")]
+            [string]$Action = "allow",
+        [Parameter (Mandatory=$false)]
+        [ValidateSet("inbound","outbound", "intra")]
+            [string]$Direction
+    )
+
+    begin {}
+    
+    process {    
+        [System.XML.XMLDocument]$xmlDoc = New-Object System.XML.XMLDocument
+        [System.XML.XMLElement]$xmlRoot = $xmlDoc.CreateElement("actionsByCategory")
+         $xmlDoc.appendChild($xmlRoot) | out-null
+            Add-XmlElement -xmlRoot $xmlRoot -xmlElementName "category" -xmlElementText $Category
+            Add-XmlElement -xmlRoot $xmlRoot -xmlElementName "action"
+            (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlroot -query "//actionsByCategory/action").SetAttribute("class", "firewallSecurityAction")
+                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action") -xmlElementName "name" -xmlElementText $Name
+                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action") -xmlElementName "description" -xmlElementText $Description
+                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action") -xmlElementName "category" -xmlElementText $Category
+                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action") -xmlElementName "isEnabled" -xmlElementText $Enabled
+                
+
+
+                if ( $SecurityGroup ) {
+                       $count = 1
+                       foreach ( $Member in $SecurityGroup) {
+
+                            Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action") -xmlElementName "secondarySecurityGroup"
+                                   
+                            #This is probably not safe - need to review all possible input types to confirm.
+                            if ($Member -is [System.Xml.XmlElement] ) {
+                                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action/secondarySecurityGroup[$count]") -xmlElementName "objectId" -xmlElementText $member.objectId
+                            } else {
+                                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action/secondarySecurityGroup[$count]") -xmlElementName "objectId" -xmlElementText $member.objectId
+                      
+                            }
+                            $count++
+                        }
+                }
+    
+                Add-XmlElement -xmlRoot (Invoke-XpathQuery -QueryMethod SelectSingleNode -Node $xmlRoot -query "//action")  -xmlElementName "applications"
```

Comment:
This line needs to be executed only if a service is defined. If no service is defined (i.e. Any) then it still creates the element but with no applications, which causes the rule to be marked as "invalidApplications = true".